src/frontend/constraint_system.rs

```diff
@@ -0,0 +1,432 @@
+  where
+    F: FnOnce() -> Result<Scalar, SynthesisError>,
+    A: FnOnce() -> AR,
+    AR: Into<String>;
+
+  /// Enforce that `A` * `B` = `C`. The `annotation` function is invoked in testing contexts
+  /// in order to derive a unique name for the constraint in the current namespace.
+  fn enforce<A, AR, LA, LB, LC>(&mut self, annotation: A, a: LA, b: LB, c: LC)
+  where
+    A: FnOnce() -> AR,
+    AR: Into<String>,
+    LA: FnOnce(LinearCombination<Scalar>) -> LinearCombination<Scalar>,
+    LB: FnOnce(LinearCombination<Scalar>) -> LinearCombination<Scalar>,
+    LC: FnOnce(LinearCombination<Scalar>) -> LinearCombination<Scalar>;
+
+  /// Create a new (sub)namespace and enter into it. Not intended
+  /// for downstream use; use `namespace` instead.
+  fn push_namespace<NR, N>(&mut self, name_fn: N)
+  where
+    NR: Into<String>,
+    N: FnOnce() -> NR;
+
+  /// Exit out of the existing namespace. Not intended for
+  /// downstream use; use `namespace` instead.
+  fn pop_namespace(&mut self);
+
+  /// Gets the "root" constraint system, bypassing the namespacing.
+  /// Not intended for downstream use; use `namespace` instead.
+  fn get_root(&mut self) -> &mut Self::Root;
+
+  /// Begin a namespace for this constraint system.
+  fn namespace<NR, N>(&mut self, name_fn: N) -> Namespace<'_, Scalar, Self::Root>
+  where
+    NR: Into<String>,
+    N: FnOnce() -> NR,
+  {
+    self.get_root().push_namespace(name_fn);
+
+    Namespace(self.get_root(), Default::default())
+  }
+
+  /// Most implementations of ConstraintSystem are not 'extensible': they won't implement a specialized
+  /// version of `extend` and should therefore also keep the default implementation of `is_extensible`
+  /// so callers which optionally make use of `extend` can know to avoid relying on it when unimplemented.
+  fn is_extensible() -> bool {
+    false
+  }
+
+  /// Extend concatenates thew  `other` constraint systems to the receiver, modifying the receiver, whose
+  /// inputs, allocated variables, and constraints will precede those of the `other` constraint system.
+  /// The primary use case for this is parallel synthesis of circuits which can be decomposed into
+  /// entirely independent sub-circuits. Each can be synthesized in its own thread, then the
+  /// original `ConstraintSystem` can be extended with each, in the same order they would have
+  /// been synthesized sequentially.
+  fn extend(&mut self, _other: &Self) {
+    unimplemented!(
+      "ConstraintSystem::extend must be implemented for types implementing ConstraintSystem"
+    );
+  }
+
+  /// Determines if the current `ConstraintSystem` instance is a witness generator.
+  /// ConstraintSystems that are witness generators need not assemble the actual constraints. Rather, they exist only
+  /// to efficiently create a witness.
+  ///
+  /// # Returns
+  ///
+  /// * `false` - By default, a `ConstraintSystem` is not a witness generator.
+  fn is_witness_generator(&self) -> bool {
```

Comment:
Do we use this method anywhere?

Reply:
Yes, most notably in the Poseidon sponge circuit's `permute_state` method.

https://github.com/Forpee/Nova/blob/54cfd7d3bceed09d141e1c78bd20f7e8a4413c33/src/provider/poseidon/sponge/circuit.rs#L127